Content.Shared/Ninja/Systems/NinjaGlovesSystem.cs

```diff
@@ -0,0 +1,325 @@
+        out EntityUid user, out EntityUid target)
+    {
+        if (args.Target != null && TryComp<NinjaGlovesComponent>(uid, out gloves)
+            && gloves.User != null
+            && !_combatMode.IsInCombatMode(gloves.User)
+            && _timing.IsFirstTimePredicted
+            && TryComp<SharedHandsComponent>(gloves.User, out var hands)
+            && hands.ActiveHandEntity == null)
+        {
+            user = gloves.User.Value;
+            target = args.Target.Value;
+
+            if (_interaction.InRangeUnobstructed(user, target))
+                return true;
+        }
+
+        gloves = null;
+        user = target = EntityUid.Invalid;
+        return false;
+    }
+
+    private void OnDoorjack(EntityUid uid, NinjaDoorjackComponent comp, InteractionAttemptEvent args)
+    {
+        if (!GloveCheck(uid, args, out var gloves, out var user, out var target))
+            return;
+
+        // only allowed to emag non-immune doors
+        if (!HasComp<DoorComponent>(target) || _tags.HasTag(target, comp.EmagImmuneTag))
+            return;
+
+        var handled = _emag.DoEmagEffect(user, target);
+        if (!handled)
+            return;
+
+        ClientPopup(Loc.GetString("ninja-doorjack-success", ("target", Identity.Entity(target, EntityManager))), user, PopupType.Medium);
+        _adminLogger.Add(LogType.Emag, LogImpact.High, $"{ToPrettyString(user):player} doorjacked {ToPrettyString(target):target}");
+    }
+
+    private void OnStun(EntityUid uid, NinjaStunComponent comp, InteractionAttemptEvent args)
+    {
+        if (!GloveCheck(uid, args, out var gloves, out var user, out var target))
+            return;
+
+        // short cooldown to prevent instant stunlocking
+        if (_useDelay.ActiveDelay(uid))
+            return;
+
+        // battery can't be predicted since it's serverside
+        if (user == target || _net.IsClient || !HasComp<StaminaComponent>(target))
+            return;
+
+        // take charge from battery
+        if (!_ninja.TryUseCharge(user, comp.StunCharge))
+        {
+            _popups.PopupEntity(Loc.GetString("ninja-no-power"), user, user);
+            return;
+        }
+
+        // not holding hands with target so insuls don't matter
+        _electrocution.TryDoElectrocution(target, uid, comp.StunDamage, comp.StunTime, false, ignoreInsulation: true);
+        _useDelay.BeginDelay(uid);
+    }
+
+    // can't predict PNBC existing so only done on server.
+    protected virtual void OnDrain(EntityUid uid, NinjaDrainComponent comp, InteractionAttemptEvent args) { }
+
+    private void OnDrainDoAfter(EntityUid uid, NinjaDrainComponent comp, DoAfterEvent args)
+    {
+        if (args.Cancelled || args.Handled || args.Args.Target == null)
+            return;
+
+        var user = args.Args.User;
+        var target = args.Args.Target.Value;
+
+        _ninja.TryDrainPower(user, comp, target);
+    }
```

Comment:
```suggestion
    private void OnDrainDoAfter(EntityUid uid, NinjaDrainComponent comp, DoAfterEvent args)
    {
        if (args.Cancelled || args.Handled || args.Args.Target == null)
            return;

        _ninja.TryDrainPower(args.Args.User, comp, args.Args.Target.Value);
    }
```

No need to set user and target as vars here, can just put them right into TryDrainPower.